activiti-cloud-runtime-bundle-service/activiti-cloud-services-runtime-bundle/activiti-cloud-services-connectors/src/main/java/org/activiti/services/connectors/channel/ServiceTaskIntegrationResultEventHandler.java

```diff
@@ -79,53 +89,103 @@ public void receive(IntegrationResult integrationResult) {
+                "The integration result for the integration context `{}` will be ignored.",
+                integrationContext.getExecutionId(),
+                integrationContext.getClientId(),
+                integrationContext.getId()
+            );
+            Command<?> cleanupCmd = getCleanupCmd(integrationContextEntity, integrationContext);
+            managementService.executeCommand(cleanupCmd);
+            return;
+        }
+
+        ExecutionEntity execution = (ExecutionEntity) executions.getFirst();
+        if (execution.getActivityId().equals(integrationContext.getClientId())) {
+            executeTriggerOrHandleError(integrationResult, execution, integrationContextEntity);
+        } else {
+            LOGGER.warn(
+                "Could not find matching activityId '{}' for integration result '{}' with executionId '{}'",
+                integrationContext.getClientId(),
+                integrationResult,
+                execution.getId()
             );
+            // If activity doesn't match, just delete the context and aggregate the event.
+            Command<?> cleanupCmd = getCleanupCmd(integrationContextEntity, integrationContext);
+            managementService.executeCommand(cleanupCmd);
+        }
+    }
+
+    private void executeTriggerOrHandleError(
+        IntegrationResult integrationResult,
+        ExecutionEntity execution,
+        IntegrationContextEntity integrationContextEntity
+    ) {
+        IntegrationContext integrationContext = integrationResult.getIntegrationContext();
 
-            managementService.executeCommand(CompositeCommand.of(commands.toArray(Command[]::new)));
+        Command<?> triggerAndCleanupCmd = CompositeCommand.of(
+            new TriggerCmd(
+                integrationContext.getExecutionId(),
+                integrationContext.getOutBoundVariables(),
+                variablesPropagator
+            ),
+            new DeleteIntegrationContextCmd(integrationContextEntity),
+            new AggregateIntegrationResultReceivedEventCmd(
+                integrationContext,
+                runtimeBundleProperties,
+                processEngineEventsAggregator
+            )
+        );
+
+        try {
+            managementService.executeCommand(triggerAndCleanupCmd);
+        } catch (Exception triggerException) {
+            LOGGER.error(
+                "Failed to trigger execution '{}' with integration result. Attempting to propagate as BPMN error.",
+                execution.getId(),
+                triggerException
+            );
+            handleTriggerFailure(triggerException, integrationContext);
         }
     }
+
+    private void handleTriggerFailure(Exception triggerException, IntegrationContext integrationContext) {
+        CloudBpmnError cloudBpmnError = new CloudBpmnError(
+            "INTEGRATION_ERROR_RECEIVED",
+            triggerException.getMessage(),
+            triggerException
+        );
```

Comment:
I'm not sure we want to trigger a cloud bpmn error in this case (the one that can be caught by an boundary error). Here something went wrong while handling the result, so I'd throw another type of exception just to get service task marked as errored.